### ISSUE TITLE
Add feature: Calendar View

### ISSUE THREAD

**Issue by xet7** [wekan/wekan#808]
_From @xet7 on January 10, 2017 17:8_

Wish originally from @TutloTutlo

It's this Trello feature:
http://help.trello.com/article/811-viewing-cards-in-a-calendar-view
http://blog.trello.com/enable-trello-calendar-mode

There's also possibility use Trello with Google Calendar:
http://help.trello.com/article/829-using-trello-with-google-calendar

This will include CalDav support.
Resource: [Introducing CalDav](http://blogs.nologin.es/rickyepoderi/index.php?/archives/14-Introducing-CalDAV-Part-I.html)

In Sandstorm Wekan could work with some Sandstorm calendar app instead.

_Copied from original issue: wefork/wekan#62_

**Comment by phillc73** [comment]
Related to #424 

I think CalDAV support would provide the most flexibility. 

Resource: [Introducing CalDav](http://blogs.nologin.es/rickyepoderi/index.php?/archives/14-Introducing-CalDAV-Part-I.html)

**Comment by NoodleBB** [comment]
Calendar view is missing. :+1:

**Comment by steven-tib** [comment]
+1 for the calendar view! I just discovered it on trello and would be awesome to have it in Wekan (making it even greater!!!) 
It would open new perspectives for the use of wekan in two current projects.

**Comment by NoodleBB** [comment]
> It would open new perspectives for the use...
It's the main reason, why i can't use wekan in reallife. 😞

**Comment by pmatts** [comment]
Easier than webdav would be to maintain an ics file.  This could then be subscribed to from a multitude of clients (including android via icsdroid).

This would be the most useful implementation for me anyways.

**Comment by Creanimo** [comment]
+1
I urgently need a calendar view as well. I think a calendar (especially with drag and drop) could turn wekan into a very effective project management software.

I am quite new to github and I saw that sometimes people put a bounty on implementing features. Is this a possibility here?

**Comment by xet7** [comment]
@Creanimo 

Yes it's possible to put bounty:
https://www.bountysource.com/teams/wekan

**Comment by xet7** [comment]
@Creanimo

Yes it's possible to put bounty:
https://www.bountysource.com/teams/wekanteam

**Comment by xet7** [comment]
There's some setup of bountysource still going on, so please wait until I have configured it correctly.

**Comment by xet7** [comment]
https://github.com/wekan/wekan/wiki/FAQ#will-my-feature-request-be-implemented

**Comment by killua-eu** [comment]
Also #1101 relevant to this issue.

**Comment by xet7** [comment]
@Creanimo 

Ok it's now possible to try to add bounty to:
https://www.bountysource.com/teams/wekan

Please tell me does it work Ok.

**Comment by killua-eu** [comment]
It works ok. If there's a dev to pick this up, I'll follow through and throw in a major $10 bountysource to verify things right to the end.

**Comment by xet7** [comment]
@killua-eu 

Yes I (or somebody else in Wekan team) will implement those features that have bounties.
Having a bounty raises priority a lot.

Please also add description how calendar view feature should work.

**Comment by xet7** [comment]
I did also add Commercial Support to wiki, if somebody is interested in that:
https://github.com/wekan/wekan/wiki

**Comment by killua-eu** [comment]
Hey @xet7 , a complete time/scheduling implementation would in my opinion consist of:

- **trello inspired calendar view https://blog.trello.com/enable-trello-calendar-mode** consisting of
   - [ ] weekly/monthly views, each with all cards stacked
   - [ ] cards draggable and filterable (changes the dates) on cal view
   - [ ] since we have both, differentiate between start and due date - not sure how about cards having both start-date and due-date defined, these would normally span multiple days in calendar apps, but that means we'll have to figure out a meaningful way to show that
   - [ ] contrary to trello, i'd consider reserving 1/n-th of the right side of the screen for tasks with no due date (so a simple todo list with items ordered according to the date they were last modified or first created)
   - [ ] caldav support (with authentication/autorization)
   - [ ] caldav url for the whole board
   - [ ] caldav url for a single user and all boards he/she participates
   - [ ] cadav url for a part of a board identified by .... (tag, user, ...) 
   - [ ] notification support (send notification 1hr, 3hrs, 1day, 2days, 1week ... ahead of due date)
- **change checklists and/or checklists items into new cards, move card to another board** Whenever you deal with using a wekan-like tool to do some time-constraint scheduling, i.e. "Build sand castle (due on 20/10/2017)" and the task becomes complex **over time**, you can end up having almost a whole board on a single card. Right now, the only way to deal with this situation is to copy out things manually.
- **odoo inspired events support #1101:** Often when something has a due date and multiple people are assigned, you have some checkpoints/events along the way "Build sand castle (due on 20/10/2017)" i.e. 
   - initial meeting (1/9/2017) in new york (location can be important too)
   - sand delivery (15/9/2017)
   - check if all's good and it seems we will finish on time (15/10/2017)
   - send out castle opening invites (16/10/2017)
that have to be recorded somewhere/somehow and eventually should show up on the calendar. also these events should be able to send notifications and should show up in the calendar view. We could bind these event data to checklist items too if the above point gets implemented.
- bidirectional caldav support (i.e. "bind" a google calendar onto a board/lane and display events)
- having teams implemented and seeing an overview of who of the members is free/unutilized (including data anonymization - if a team member is busy on tuesday on a board i'm not participating in, i should see that he's busy but shouldnt know why)
- add card dependence (card x depends on card y)
- use dependence and calendar data to get a gantt chart
- list of tasks from all boards (can be already done via api)

Hope this paints the picture. I took your initial description of this issue and expanded it slightly taking into account what might be a problem or what usecases come to my mind. If you'd like me to add more details just let me know.

**Comment by killua-eu** [comment]
And the bounty is here: https://www.bountysource.com/issues/41780513-add-feature-calendar-view , @xet7 , you're free to test the rest.

**Comment by xet7** [comment]
@killua-eu 

It seems that super-productivity does not have calendar yet (see issue above) but has many other nice features.

For calendar, there are some packages for meteor:
https://atmospherejs.com/packages/calendar

For example:
https://atmospherejs.com/fullcalendar/fullcalendar
https://fullcalendar.io

I have not checked how many of required features it has.

**Comment by killua-eu** [comment]
Firstly I'd add 

- https://github.com/Serhioromano/bootstrap-calendar
- https://dhtmlx.com/docs/products/demoApps/

to the possible libraries to use. dhtmlx and fullcalendar are the two most prominently used ones imho, so that would be probably the best way to go. AFAIK fullcalendar doesnt have caldav support, but I guess none will have that. dhtmlx has a gplv2 version and i think is more polished, but i never used it in a project.

**Comment by Creanimo** [comment]
@killua-eu That is a very complete and helpful list you posted. It would be amazing if even a fraction of these features could be added!

Is there any dev here who wants to tackle this soon/next? I would rather hire someone who is already familiar with wekan.

**Comment by killua-eu** [comment]
@Creanimo, I developed an ERP that had all this and more included, so I'll be having lots of ideas (see my other issues in wekan), unfortunately is my javascript quite weak, otherwise I'd hack on this myself. So at least, I try to help this way. BTW if you like this issue, you'll probably like #1160 as well. If you'd join me in  adding a bounty on this on bountysource (see above), I'll top up my bounty too - sofar I testpledged only $10. If you wanna go faster, please note that @xet7 also wrote up an article on commercial devrequests too on the wiki. I'll continue to provide any guidance/architecture tips to anyone who may find them useful.

**Comment by xet7** [comment]
I started implementing this yesterday 2017-08-11.

At first I'm targeting for enough functionality to be usable.

# STATUS 2017-08-12

## DONE

- add fullcalendar plugin to Wekan https://atmospherejs.com/rzymek/fullcalendar
- tested Calendar button on Card menu without data, and later removed it from Card menu
- added Calendar button to Wekan top right bar, in similar place like in Trello. Button currenly only opens small popup, it needs to be resized.

## TODO

### 1) Create standalone demo outside Wekan to combine these fullcalendar demos:

LOCALES:  https://fullcalendar.io/js/fullcalendar-3.4.0/demos/locales.html
TIMEZONE: https://fullcalendar.io/js/fullcalendar-3.4.0/demos/timezones.html
TIME RANGES: https://fullcalendar.io/js/fullcalendar-3.4.0/demos/selectable.html (possible to enlarge or shorten time range by dragging)
GOOGLE CALENDAR: https://fullcalendar.io/js/fullcalendar-3.4.0/demos/gcal.html (create similar URLs inside Wekan what fullcalendar shows)
EXTERNAL EVENTS: https://fullcalendar.io/js/fullcalendar-3.4.0/demos/external-dragging.html (for dragging Wekan cards to board)

### 2) Create gcal/caldav URLs inside Wekan to show start and due dates.

Technically:
- start and due dates are at MongoDB database
- generate gcal/caldav URLs live, similar to displaying card changes immediately in Wekan, or change to start/duedate triggering update
- add gcal/caldav authentication

### 3) Calendar button shows screen that shows data from gcal/caldav URLs:

- Top: Language and timezone selection - preselected according to Wekan language (or browser setting)
- Top: Change to previous or next month
- Left: Calendar display.
- Left: List of those that only have start time, no due date. On calendar those could be shown as having same start time and due date, and maybe using different color.
- Right Top: Dropdowns of lists and their cards, selected card can then be dragged to calendar using external events above.
- Right Bottom: Filtering by user.

## FEEDBACK WELCOME

If anyone has improvements to this design, please add comment to this issue.

**Comment by killua-eu** [comment]
Hey @xet7 , thanks for the update, sorry for the late reply. As for now, I'd probably generate gcal/caldav URLs live, as long as this feature is in flux, triggered updates might just leave bits here and there with wrong data. As for point 3, this hits a bit the limits of my imagination, should you have a screenshot at hand, the picture surely would paint a 1000 words.

**Comment by Creanimo** [comment]
I am very excited for this!

Will the calendar be able to show the cards of selected and all boards?

**Comment by xet7** [comment]
@Creanimo 

If possible, yes. I will update this issue as I get more progress.

**Comment by Creanimo** [comment]
Is there an ETA yet? O:-)

**Comment by xet7** [comment]
@Creanimo

:D I really should try to continue this. When I have had free time for Wekan, there has been so many pull requests to merge, bugfixes to do, releases to do, etc so I probably forgot about this, so thanks for reminding :)

**Comment by xet7** [comment]
I updated [roadmap](https://github.com/wekan/wekan/wiki/Roadmap) :)

**Comment by lichtamberg** [comment]
Yeah, please continue this :) I really need this feature :P

**Comment by xet7** [comment]
Related Python calendar implementation by @LukasGasior1

https://discourse.wekan.io/t/sharing-my-calendar-integration-plugin/601

https://github.com/LukasGasior1/wekan-ical-server

**Comment by Creanimo** [comment]
Do you have an ETA for us? My pledge at bounty source has a 3 month deadline on it, but I will re-pledge should the deadline pass.

**Comment by xet7** [comment]
@Creanimo 

I was not aware of the deadline. Where I can see deadlines of all bounties?

**Comment by killua-eu** [comment]
I think that on bounties above 150 Eur, there's a default 3 month deadline. The list of supporters and the date they pledged is here https://www.bountysource.com/issues/41780513-add-feature-calendar-view/backers

**Comment by xet7** [comment]
@Creanimo 

I need to update Roadmap order, because I need to do in this order:

1) search
2) copy + move cards to other boards
3) calendar

I don't have ETA, but I try to do more Wekan and less other work.

**Comment by xet7** [comment]
Note to myself of these features wishes:

- modify the start and due dates directly in the calendar
- filter the cards with selected tags or user

**Comment by xet7** [comment]
I reorganized [Roadmap](https://github.com/wekan/wekan/wiki/Roadmap) and started to think that it's probably good idea to do those first that are currently in progress, before starting anything else.

**Comment by lichtamberg** [comment]
Is there any date where this will (maybe) available? I would really love to rollout wekan in our company, but our chief says without proper calender this is a show stopper :/ Just want to ask.. Thank you very much for this cool product!

**Comment by xet7** [comment]
@lichtamberg 

I plan to have calendar feature working this year, if possible.

Also note that Let's Encrypt wildcard certificates will be available at January next year:
https://letsencrypt.org/2017/07/06/wildcard-certificates-coming-jan-2018.html
There is already plans to have it in Sandstorm, it makes using Sandstorm a lot easier.

For companies I recommend commercial support of Wekan Team, if you have some budget:
http://www.xet7.org/wekan
It really helps to speed up Wekan development, encourages contributors etc.

**Comment by Creanimo** [comment]
Just want to say how much I appreciate your hard work and that I hope that more people will use your commercial services and bountysource to support the project. Is great that wekan is still alive and developing. Side note about SSL, I am using Wekan in cloudron.io which already provides Let's Encrypt certificates for all apps. Not sure how that works technically but maybe you can benefit from some source code that is used for that.

Back to topic: I have been using Zenkit which has (among many other) a calendar view (and custom fields which are also on the roadmap). It's amazing how that takes a kanban board to a very capable project management. I think with a calendar view we can promote Wekan as a project management software, which might get you more commercial users. At least for my studio I try to use as little apps as possible and merging my brainstorming and collection tool (Wekan) with project management will streamline things a lot. Thanks again! I am excited for the future of Wekan!

**Comment by xet7** [comment]
@Creanimo 

Thanks!

There is [LDAP in Cloudron Wekan](https://github.com/wekan/wekan/issues/119#issuecomment-287597821) but it's not integrated yet.

**Comment by xet7** [comment]
@Creanimo 

For Let's Encrypt certificates, current self-hosting Wekan alternatives are using in front of Wekan:
- [Caddy](https://github.com/wekan/wekan/wiki/Let's-Encrypt-and-Google-Auth)
- [Nginx](https://github.com/wekan/wekan/wiki/Nginx-Webserver-Config) with [Certbot](https://certbot.eff.org/)

If you want longer time between renewing SSL certs than 3 months, you can use free account at [CloudFlare](https://www.cloudflare.com/) and there click Crypto/Origin Certificates/Create certificate that is used to encrypt traffic between your server's Nginx and CloudFlare.

**Comment by xet7** [comment]
Sorry, calendar has not progressed yet.

Here are some excuses what happened instead:
- Spending Chrismas with family etc
- Translating Sandstorm to Finnish
- Translating ReactOS Start button/menu and Taskbar to Finnish
- Trying to get Wekan upgraded Meteor 1.6 and Node 8.8.1 with no success
- Trying to get Wekan working on ARM64 yesterday on server I got from CNCF, stuck on fibers. 
- Merging some pull requests, but did not make release
- Tried to fix double dash bug on snap, but because Ubuntu is importing packages from Debian, build queue is progressing very slowly, I have not checked has is built yet
- Having some urgent work for my clients
- Did blog post about 2 years of Wekan

Well, I'll try to organize my time that I could hopefully make progress with calendar...

**Comment by xet7** [comment]
Additional requirements, sponsored by FX:

When modifying due date of a card:
1) Show info about due date change in the activities panel
2) Email card members about change of due date

**Comment by xet7** [comment]
@BaobabCoder has started implementing this as minimal implementation first.

**Comment by xet7** [comment]
Steps:
1) @BaobabCoder makes minimal implementation of Calendar feature, and this issue is closed
2) @BaobabCoder gets full bounty of this feature, just like @GhassenRjab got for minimal implementation of Search feature
3) Remaining Calendar feature requests are moved to new issue

**Comment by xet7** [comment]
@BaobabCoder has not provided any status report, so for now, I'm moving this back to my todo list.

**Comment by xet7** [comment]
@BaobabCoder came back, and is now continuing Calendar :)

**Comment by xet7** [comment]
Correction: @BaobabCoder said that he makes basic first version of calendar, adds it as pull request, and gives bounty to me, so that I can continue adding features and updates to calendar, and to support my development of Wekan, because @BaobabCoder currently has his full time job, and I only have some part-time job, doing Wekan etc. Thanks to @BaobabCoder !

**Comment by d3dbit** [comment]
IMHO, calendar view is a needed feature to improve wekan app. Thanks for continuing developing it.

**Comment by xet7** [comment]
Related: Calendar versioning https://calver.org

**Comment by xet7** [comment]
Calendar in progress at https://github.com/wekan/wekan/pull/1728

**Comment by jinnko** [comment]
Bidirectional support would be the killer feature here.  CalDAV task support is basically universal and has many interfaces such as macOS Reminders, Thunderbird, etc., yet there's no Kanban type integration.  @killua-eu mentioned this in the list of desired features.  Are you considering bi-directional support in scope for this issue?

**Comment by nickmdowson** [comment]
Just installed Wekan and experimenting with it - am I right in thinking the calendar view doesn't do anything yet? (and is there any way currently to filter by due dates, etc?) Thanks!

**Comment by sfahrenholz** [comment]
@nickmdowson 
the calender view exists already, on the main menu (right side) is the function to switch between Swimlanes, Lists,, Calender.

![image](https://user-images.githubusercontent.com/6977657/46631088-a5866e00-cb46-11e8-9e11-722d8cb4cd2b.png)

**Comment by frakman1** [comment]
@cannycartographer I tried adding start and end dates to my tasks but  still nothing shows up in any of the calendar views. What am I missing? I am using Wekan version 1.78.0 and Node version 8.12.0 on my Mac using Docker (dockerbunker). See screenshots below:

Job starts and stops are Nov 21st to Nov 24th:
![image](https://user-images.githubusercontent.com/5826484/48950111-1c65a380-ef08-11e8-9cc5-dfee3aba6130.png)

Nothing displayed on those dates in the calendar for the month of November:
![image](https://user-images.githubusercontent.com/5826484/48950185-52a32300-ef08-11e8-9ec5-581a7ca09cfd.png)

**Comment by xet7** [comment]
@frakman1 

Calendar Empty is issue #1998 . This issue is closed.

**Comment by frakman1** [comment]
@xet7 Thank you for the link. I don't understand why it is closed? I have both the start and end date set and yet I don't see any item in the calendar. Are you saying it is a bug that won't be fixed?

**Comment by xet7** [comment]
@frakman1 

Calendar View has been added to Wekan, and it did work in Wekan when it was originally added.

Calendar being broken, or adding new feature to calendar, is a new issue. It is not related to old issue, and is not continued at old issue. Old issues are closed, when they are done.

I do often combine and move issues. Wekan bugs will be fixed.

So, please continue at issue #1998 . Do not continue at this old issue.

**Comment by frakman1** [comment]
I understand now. Thank you for clearing that up.
